front/components/assistant_builder/ActionsScreen.tsx

```diff
@@ -578,7 +654,35 @@ function ActionConfigEditor({
   setEdited: (edited: boolean) => void;
   description: string;
   onDescriptionChange: (v: string) => void;
-}) {
+};
+
+function ActionConfigEditor({
+  owner,
+  action,
+  vaultsUsedInActions,
+  instructions,
+  updateAction,
+  setEdited,
+  description,
+  onDescriptionChange,
+}: ActionConfigEditorProps) {
+  const { vaults } = useContext(AssistantBuilderContext);
+
+  // Only allow one vault across all actions
+  const allowedVaults = useMemo(() => {
+    const usedVaultsInOtherActions = vaults.reduce((acc, v) => {
+      return (vaultsUsedInActions[v.sId] ?? []).some((a) => a !== action)
+        ? acc.concat(v)
+        : acc;
+    }, [] as VaultType[]);
+
+    return vaults.filter(
+      (vault) =>
+        usedVaultsInOtherActions.length == 0 ||
+        usedVaultsInOtherActions.find((v) => v.sId === vault.sId)
+    );
+  }, [action, vaults, vaultsUsedInActions]);
```

Comment:
Pretty hard to read IMO, may I suggest something like:
```suggestion
const allowedVaults = useMemo(() => {
  const isVaultUsedInOtherActions = (vault: VaultType) => {
    const actionsUsingVault = vaultsUsedInActions[vault.sId] ?? [];
    return actionsUsingVault.some((a) => a.name !== action.name);
  };

  const usedVaultsInOtherActions = vaults.filter(isVaultUsedInOtherActions);

  if (usedVaultsInOtherActions.length === 0) {
    return vaults;
  }

  return vaults.filter((vault) =>
    usedVaultsInOtherActions.some((v) => v.sId === vault.sId)
  );
}, [action, vaults, vaultsUsedInActions]);
```

Reply:
Done (with action vs action.name for the reason explained above)